lib/cosmos/gui/choosers/integer_chooser.rb

```diff
@@ -9,72 +9,59 @@
 # attribution addendums as found in the LICENSE.txt
 
 require 'cosmos'
+require 'cosmos/gui/choosers/value_chooser'
 
 module Cosmos
 
-  class IntegerChooser < Qt::Widget
+  class IntegerChooserIntValidator < Qt::IntValidator
+    def initialize(*args)
+      super(*args)
+    end
 
+    def fixup(input)
+      begin
+        value = input.to_i
+        if value < bottom()
+          # Handle less than bottom
+          parent().setText(bottom().to_s)
+        elsif value > top()
+          # Handle greater than top
+          parent().setText(top().to_s)
+        elsif input != value.to_s
+          # Handle poorly formatted (only known case is float given as starting value)
+          parent().setText(value.to_s)
+        end
+      rescue Exception => err
+        # Oh well no fixup
+      end
+    end
+  end
+
+  class IntegerChooser < ValueChooser
     # Callback called when the value changes
     attr_accessor :sel_command_callback
 
-    def initialize(
-      parent, label_text, initial_value,
-      minimum_value = nil, maximum_value = nil, field_width = 20, fill = false
-    )
-      super(parent)
+    def initialize(parent, label_text, initial_value,
+                   minimum_value = nil, maximum_value = nil,
+                   field_width = 20, fill = false)
+      super(parent, label_text, initial_value, field_width, fill)
       @minimum_value = minimum_value
       @maximum_value = maximum_value
 
-      layout = Qt::HBoxLayout.new(self)
-      layout.setContentsMargins(0,0,0,0)
-
-      @integer_label = Qt::Label.new(label_text)
-      @integer_label.setSizePolicy(Qt::SizePolicy::Fixed, Qt::SizePolicy::Fixed) if fill
-      layout.addWidget(@integer_label)
-      layout.addStretch() unless fill
-      @integer_value = Qt::LineEdit.new(initial_value.to_s)
-      @integer_value.setMinimumWidth(field_width)
-      if minimum_value or maximum_value
-        validator = Qt::IntValidator.new(@integer_value)
-        validator.setBottom(minimum_value) if minimum_value
-        validator.setTop(maximum_value) if maximum_value
-        @integer_value.setValidator(validator)
-      end
-      @callback_in_progress = false
-      @integer_value.connect(SIGNAL('editingFinished()')) do
-        unless @callback_in_progress # Prevent double fire on loss of focus
-          begin
-            @callback_in_progress = true
-            @sel_command_callback.call(string(), value()) if @sel_command_callback
-          ensure
-            @callback_in_progress = false
-          end
-        end
-      end
-      layout.addWidget(@integer_value)
-      setLayout(layout)
-
-      @sel_command_callback = nil
-    end
-
-    # Returns the value as a string
-    def string
-      @integer_value.text
+      validator = IntegerChooserIntValidator.new(@value)
+      validator.setBottom(minimum_value) if minimum_value
+      validator.setTop(maximum_value) if maximum_value
+      @value.setValidator(validator)
```

Comment:
We always want to set the validator even if there isn't a min or max because it still provides validation.